### ISSUE TITLE
How to detect client disconnection during HTTP streaming request (chunked encoding)?

### ISSUE THREAD

**Issue by bryant-ferguson** [civetweb/civetweb#1090]
I have an application that responds to certain paths with an HTTP stream. 
i.e. I use these headers:
```
Connection: keep-alive
Transfer-Encoding: chunked
```
and send data with `mg_send_chunk`.

When testing, I typically use curl as my client and will Ctrl-C to disconnect.

The only way that I currently know how to detect that a client has disappeared is that `mg_send_chunk` eventually replies with a `-1` return code. Typically, this happens after I try to send 2 or 3 chunks. The first 1 - 2 chunks return successfully from the call to `mg_send_chunk` with the length of the data sent.

So, I have two questions:
1) Why is success returned at all after I disconnect the client?
2) Is there a way to detect that the client disconnected (by polling or callback) without sending data to the client?

Thanks. I hope this is clear; I don't consider myself an expert in this area.

**Comment by bel2125** [comment]
Sorry for the delay, I was on extended summer vacation.

> Why is success returned at all after I disconnect the client?

TCP/IP communication is asynchronous on multiple levels.
A successful "send" call will tell you nothing more than that the network stack in the sending partner accepted the data you gave it. It does not mean that this data is already sent ... or will be sent ... ever. And it definitely does not mean it was received at the other side.
The network stack will try asynchronously to transport the data to the receiving network stack, probably through multiple intermediates.
Eventually the receiving side will send a "received" (acknowledged) message.
Only if the sending side has sent too many unacknowledged data, it will eventually no longer accept additional "send" calls and return an error. That may take many minutes, depending on the network, configurations, ...

Even if the sending side already closed the connection, there may be many packages "in transition". Still on the sending stack, in some intermediate, maybe even already at the receiving network stack, but not yet received by the application.

Your receiving application will receive all this data, until there is no more to receive - then you will see the error.

**Comment by bel2125** [comment]
> Is there a way to detect that the client disconnected (by polling or callback) without sending data to the client?

There is a timeout for open connections you can set.

But this really depends on your specific needs.

Usually HTTP will close the connection after receiving a reply for the request anyway (yes, there is keep-alive - but it is always valid to close the connection, even for an intermediate).
If you need to have a connection open, although you are not sending data now, you can use websockets.
They also have a timeout, so you need to send an "empty" message every now and then. This is already implemented in a standardized way using "websocket ping pong".

However, I probably could write a novel as reply to this answer if I want to cover everything. 
I could give you an answer more to the point if I know your specific requirements.
Or I could give you a general advice to build a HTTP application in a way that it does not need to know this.

**Comment by bryant-ferguson** [comment]
@bel2125 I'll try to describe my usage with more detail here. I really appreciate you taking the time to respond. I don't necessarily understand all of the details at play here nor do I completely understand everything you wrote above, so forgive my ignorance. Also, if this gets to be too complicated, we can either drop it all together, or take this discussion elsewhere. Thanks.

We have implemented a public API. Requests arriving at one particular path (we call this our "http stream") respond with `keep-alive` in the header and delivers "events" as long as the connection stays open. We also send "\r\n" as a keep-alive every 30 seconds if no events are already being sent. Unless I misunderstand how they work, websockets are not an option since that would be a change to the public API. We could add support for a websocket stream, but that's outside the scope of this discussion... I think.

Recently, we added support for more "http streams" than usual (we used to limit the number of streams 5 and upon a 6th connection, we would close the oldest from the server and accept the newest). I started doing some testing to examine the effects on the system when we run out of worker threads, so I artificially limited workers to 2.

This is when I noticed that streams are not cleaned up immediately when the client disconnects. With 2 worker threads, I made two requests to the http stream path and left them open. Then I made a 3rd request to a different path that does not use "keep-alive" in the header (just a typical request that responds and closes the connection immediately). As expected that 3rd request sat un-serviced as it was waiting on a worker thread. Then, I disconnected one of the http streams from the client side. I expected that the worker thread would free up immediately and service the 3rd request, but it took over a minute to respond. This is partially because the 2 "http-streams" were not actually delivering events, only keep-alive messages every 30 seconds. I added some print statements to the code and ran this experiment several times and every time, it took exactly 3 keep-alives before civetweb realized the connection was dead and returned a failure, triggering the server to clean-up the thread.

So, is there anything we can do to clean-up these client-closed connections faster without changing the rate of keep-alive messages or events. It sounds as though perhaps the answer is no. Note that I think we are closing the connection on the client side "gracefully" (sending Ctrl-C to curl). So I figure there may be something in the protocol to identify that graceful disconnect.   If there is nothing we can do, that's also useful information. Thanks!

**Comment by bel2125** [comment]
By "http streams" you don't mean "streams" in the same sense as HTTP/2 does. I'm noting this just that we don't run into confusion because of this name. But HTTP/2 would not help here either, since it's a TCP level topic, and that is the same. TCP is the protocol level on which both HTTP/1.x and HTTP/2 are based on.

You can reduce the timeout from 30 seconds to a lower value.
This is a configuration parameter.

I thought about it for a while, but I still come to the conclusion this is how TCP works, and what you observe is a "feature" of this protocol. Unwanted in your case, but nevertheless a feature designed into TCP.

**Comment by bryant-ferguson** [comment]
I don't believe it's an HTTP/2 thing. It's just a GET request from the client that stays open from `Connection: keep-alive` being specified in the header. 

I believe it's modeled after the Twitch streaming API:
https://developer.twitter.com/en/docs/tutorials/consuming-streaming-data

Regardless, sounds like there's not much to do here other than to try setting the timeout.

Which timeout is that? `linger_timeout_ms`?

**Comment by bel2125** [comment]
It is definitely NOT a HTTP/2 issue. I only mentioned because of the word "stream" - we must not intermix "HTTP/2 stream" and "TCP stream". This topic is related to "TCP stream".

I meant the general request timeout:

https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#request_timeout_ms-30000

You can test to set the linger timeout, but I don't think this will help in your case:

https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#linger_timeout_ms

If you use keep-alive, there is also another timeout, but from your issue description, I don't think this will help you either:

https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#keep_alive_timeout_ms-500-or-0
